lightstep/sdk/metric/data/data.go

```diff
@@ -0,0 +1,140 @@
+	// push-based exporters that perform sequential collection.
+	Metrics struct {
+		// Resource is the MeterProvider's configured Resource.
+		Resource *resource.Resource
+
+		// Scopes is a slice of metric data, one per Meter.
+		Scopes []Scope
+	}
+
+	// Scope is the result of a single collection for a single Meter.
+	//
+	// See the comments on Metrics about re-use of slices in this struct.
+	Scope struct {
+		// Library describes the instrumentation scope.
+		Library instrumentation.Library
+
+		// Instruments is a slice of metric data, one per Instrument
+		// in the scope.
+		Instruments []Instrument
+	}
+
+	// Instrument is the result of a single collection for a single Instrument.
+	//
+	// See the comments on Metrics about re-use of slices in this struct.
+	Instrument struct {
+		// Descriptor describes an instrument created through a View,
+		// including name, unit, description, instrument and number kinds.
+		Descriptor sdkinstrument.Descriptor
+
+		// Points is a slice of metric data, one per attribute.Set value.
+		Points []Point
+	}
+
+	// Point is a timeseries data point resulting from a single collection.
+	Point struct {
+		// Attributes are the coordinates of this series.
+		Attributes attribute.Set
+
+		// Temporality applies to instruments that have it defined
+		// (which excludes the asynchronous gauge).
+		Temporality aggregation.Temporality
+
+		// Aggregation determines the kind of data point
+		// recorded in this series.
+		Aggregation aggregation.Aggregation
+
+		// Start indicates the start of the collection
+		// interval reflected in this series, which is set
+		// according to the configured temporality.
+		Start time.Time
+
+		// End indicates the moment at which the collection
+		// was performed.
+		End time.Time
+	}
+)
+
+// Reset sets the size of every slice to zero, while maintaining
+// capacity, to support re-use.
+func (m *Metrics) Reset() {
+	resetScopes(&m.Scopes)
+}
+
+func resetScopes(ss *[]Scope) {
+	for i := range *ss {
+		(*ss)[i].Reset()
+	}
+	(*ss) = (*ss)[0:0:cap((*ss))]
+}
+
+// Reset sets the size of every slice to zero, while maintaining
+// capacity, to support re-use.
+func (s *Scope) Reset() {
+	resetInstruments(&s.Instruments)
+}
+
+func resetInstruments(is *[]Instrument) {
+	for i := range *is {
+		(*is)[i].Reset()
+	}
+	(*is) = (*is)[0:0:cap((*is))]
+}
+
+// Reset sets the size of every slice to zero, while maintaining
+// capacity, to support re-use.
+func (in *Instrument) Reset() {
+	resetPoints(&in.Points)
+}
+
+func resetPoints(ps *[]Point) {
+	(*ps) = (*ps)[0:0:cap((*ps))]
+}
+
+// ReallocateFrom returns the pointer to the next available slice
+// location, allowing re-use by trying to extend the length of an
+// existing slice up to its capacity before allocating new storage.
+func ReallocateFrom[T any](p *[]T) *T {
+	if len(*p) < cap(*p) {
+		(*p) = (*p)[0 : len(*p)+1 : cap(*p)]
+	} else {
+		var empty T
+		(*p) = append(*p, empty)
+	}
+	return &(*p)[len(*p)-1]
+}
```

Comment:
Yes, that could be a concern. The library takes steps to ensure that the slices are re-used in the same order each time, except for the Points array at the leaf of the structure, which is random-order due to a map iteration. This makes it so that if there's a large allocation here, at least it will be re-used by the same instrument. 

If the SDK is cumulative, the allocation will always grow and never shrink. If delta temporality is used, then if a high-cardinality event happens and then stops happening, there will be unused memory here.